### ISSUE TITLE
[Feature] System.Text.Json support

### ISSUE THREAD

**Issue by HarelM** [elastic/elasticsearch-net#4152]
Feature request:
Support `System.Text.Json` writers.
More details:
Currently NEST supports Json.Net serializers. Microsoft has added a Json library to .net core 3 and a nuget package for other frameworks.
Will it be possible to use custom writers that supports `System.Text.Json` with NEST - I know you made a great effort to remove the JSON.net dependency, I hope this is a step forward in that aspect...
https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-apis/

**Comment by russcam** [comment]
Researching a move to `System.Text.Json` is on the roadmap for the future, though we don't have a timeline for when this will be. 

> Will it be possible to use custom writers that supports System.Text.Json with NEST

You can do this with `JsonNetSerializer` and [`Nest.JsonNetSerializer` nuget package](https://www.nuget.org/packages/Nest.JsonNetSerializer) currently. It should be possible to do something similar for `System.Text.Json` too.

**Comment by Mpdreamz** [comment]
We started exploring this but hit some bugs in System.Text.Json that prevents us currently to move to `System.Text.Json`

See: https://github.com/elastic/elasticsearch-net/pull/4211

**Comment by ArcadeRenegade** [comment]
I am currently switching over my project from Newtonsoft.JSON to System.Text.Json. Being able to use System.Text.Json for ElasticSearch.NET would be a nice to have as I will be able to remove Newtonsoft.JSON from my project entirely. +1

**Comment by Mpdreamz** [comment]
Once this lands: https://github.com/dotnet/runtime/pull/32669
And makes its way into a prelease we can continue our open PR #4211

**Comment by HarelM** [comment]
Did someone here managed to create a `IElasticsearchSerializer` implementation using `System.Text.Json` by any chance?
I'd like to avoid doing something someone already did...

**Comment by Mpdreamz** [comment]
There is one in https://github.com/elastic/elasticsearch-net/pull/4679 @HarelM 

Our target version to switch over OOTB is `8.x` although we are hopeful to start offering an `System.Text.Json` `IElasticsearchSerializer` as a separate package for the `7.x` range of the clients as a non breaking feature. 

That said there are several kinks to work out still on that PR as well as doing more performance testing work.

**Comment by HarelM** [comment]
@Mpdreamz thanks for the info!
Please note that this implementation does not allow adding custom converters as it doesn't expose the `JsonSerializerOptions` inner objects.
I might be missing out on something though...

**Comment by HarelM** [comment]
Also please consider the following bug I found in the existing serializer when implementing the new one: #4791
I thought about using this seriazlier since I had the following bug:
https://github.com/IsraelHikingMap/Site/issues/1247
The solution was to fix the current serializer since I needed to change too much of my code in order to migrate to STJ. Never-the-less, one can learn from my experience :-)

**Comment by meywd** [comment]
Hi any update on this, I see that [#4679](https://github.com/elastic/elasticsearch-net/issues/4791) is merged so is the support for System.Text.Json complete?

**Comment by stevejgordon** [comment]
We are actively working on this for v8 (high-level). Already available for Low-Level.

**Comment by meywd** [comment]
@stevejgordon Thank you for the update.

**Comment by meywd** [comment]
@stevejgordon since the high level client uses the lower level one, shouldn't this be already available?

**Comment by stevejgordon** [comment]
@meywd While the high-level dispatches through the low level, it manages it's own serialization for the more complex strongly-typed classes. Today it uses an customised Utf8Json serializer.

**Comment by stevejgordon** [comment]
We have switched to `System.Text.Json` for `8.0.0-alpha.1` as the default fore request/response and source serialization.

**Comment by HarelM** [comment]
@Mpdreamz does this systemtextjsonserializer exposed in version 7 in any way?
Can I use STJ without upgrading to version 8?
I'm revisiting this issue after NTS had added support for STJ in their geojson serialization and deserialization.

**Comment by stevejgordon** [comment]
@HarelM STJ applies only to v8, but you could create your own [custom source serializer](https://www.elastic.co/guide/en/elasticsearch/client/net-api/7.17/custom-serialization.html) that uses STJ for your types.

**Comment by HarelM** [comment]
Can I use version 8 client with version 7 database?
It's a lot of code to add just for supporting STJ...

**Comment by stevejgordon** [comment]
@HarelM No, the v8 client is not tested for compatibility with v7 and is not supported for use in that way. It shouldn't be a huge amount of code to implement the serializer interface. Our v8 source serializer would be a good starting point also, if you go that route.

**Comment by HarelM** [comment]
Yea, I saw the custom serialized code, with the extra two jsonconverters and copied them to my protect, I agree, it's not a lot of code, but also not a few simple lines. Since this code will be in my protect I need to fully understand it, which at this point in time I don't...
Thanks for the super fast response and detailed explanation! Keep up the good work!

**Comment by HarelM** [comment]
I ran the code and was able to migrate to STJ, thanks for all the help!!!
Do I have to add the exception converter and dictionary converter to my code in order for this to run as expected? I looked at the version 8 code and it's completely different - it inherits from `Serializer` and I believe I can't use that code with version 7.
I looked at the merge PR that was linked in this thread and copied from there, it had some TODOs in it, so I was wondering if there was a final version before drastically changing it?
Here's my code if it helps:
https://github.com/IsraelHikingMap/Site/blob/585bb6a8fda9f8ba95eae1da33527072fe85a1ba/IsraelHiking.DataAccess/ElasticSearchGateway.cs#L165
